File: tests/FSharpAux.Mocha.Tests/fable/src/FSharpAux.Core/Seq.js
import { exists, filter, initialize, choose, ofList, foldBack, tryPick, map, empty as empty_1, toList, iterate, length as length_1, enumerateWhile, enumerateUsing, delay, concat, initializeInfinite, take, singleton, append } from "../../fable_modules/fable-library.4.0.6/Seq.js";
import { head, tail, ofArrayWithTail, isEmpty, empty, singleton as singleton_1, cons } from "../../fable_modules/fable-library.4.0.6/List.js";
import { defaultOf, structuralHash, equals, compare, disposeSafe, getEnumerator } from "../../fable_modules/fable-library.4.0.6/Util.js";
import { HashSet } from "../../fable_modules/fable-library.4.0.6/MutableSet.js";
import { HashIdentity_Structural } from "../../fable_modules/fable-library.4.0.6/FSharp.Collections.js";
import { addToDict, tryGetValue, addToSet } from "../../fable_modules/fable-library.4.0.6/MapUtil.js";
import { intersectWith } from "../../fable_modules/fable-library.4.0.6/Set.js";
import { FSharpRef } from "../../fable_modules/fable-library.4.0.6/Types.js";
import { FSharpMap__get_Item, FSharpMap__ContainsKey, ofSeq } from "../../fable_modules/fable-library.4.0.6/Map.js";
import { value as value_1, some } from "../../fable_modules/fable-library.4.0.6/Option.js";
import { Dictionary } from "../../fable_modules/fable-library.4.0.6/MutableMap.js";
import { rangeDouble } from "../../fable_modules/fable-library.4.0.6/Range.js";
import { isInfinity } from "../../fable_modules/fable-library.4.0.6/Double.js";

export function appendSingleton(s, value) {
    return append(s, singleton(value));
}

export function consSingleton(s, value) {
    return append(singleton(value), s);
}

export function initRepeatValue(length, value) {
    return take(length, initializeInfinite((_arg) => value));
}

export function initRepeatValues(length, values) {
    return concat(take(length, initializeInfinite((_arg) => values)));
}

export function groupAfter(f, input) {
    const group = (en_mut, cont_mut, acc_mut, c_mut) => {
        group:
        while (true) {
            const en = en_mut, cont = cont_mut, acc = acc_mut, c = c_mut;
            if (!f(en["System.Collections.Generic.IEnumerator`1.get_Current"]()) && en["System.Collections.IEnumerator.MoveNext"]()) {
                en_mut = en;
                cont_mut = ((l) => cont(cons(c, l)));
                acc_mut = acc;
                c_mut = en["System.Collections.Generic.IEnumerator`1.get_Current"]();
                continue group;
            }
            else {
                return cont(singleton_1(c));
            }
            break;
        }
    };
    return delay(() => enumerateUsing(getEnumerator(input), (en_1) => enumerateWhile(() => en_1["System.Collections.IEnumerator.MoveNext"](), delay(() => singleton(group(en_1, (x) => x, empty(), en_1["System.Collections.Generic.IEnumerator`1.get_Current"]()))))));
}

export function groupWhen(f, input) {
    let matchValue_2, t_2, h_2;
    const en = getEnumerator(input);
    try {
        let firstCase = false;
        const loop = (cont_mut) => {
            loop:
            while (true) {
                const cont = cont_mut;
                if (en["System.Collections.IEnumerator.MoveNext"]()) {
                    if (f(en["System.Collections.Generic.IEnumerator`1.get_Current"]())) {
                        const temp = en["System.Collections.Generic.IEnumerator`1.get_Current"]();
                        cont_mut = ((y) => {
                            let firstCase_1;
                            return cont((firstCase_1 = firstCase, isEmpty(y) ? ((firstCase = true, singleton_1(singleton_1(temp)))) : (firstCase_1 ? ((firstCase = false, ofArrayWithTail([empty(), singleton_1(temp)], y))) : ofArrayWithTail([empty(), cons(temp, head(y))], tail(y)))));
                        });
                        continue loop;
                    }
                    else {
                        const temp_1 = en["System.Collections.Generic.IEnumerator`1.get_Current"]();
                        cont_mut = ((y_1) => {
                            let firstCase_2;
                            return cont((firstCase_2 = firstCase, isEmpty(y_1) ? singleton_1(singleton_1(temp_1)) : (firstCase_2 ? ((firstCase = false, cons(singleton_1(temp_1), y_1))) : cons(cons(temp_1, head(y_1)), tail(y_1)))));
                        });
                        continue loop;
                    }
                }
                else {
                    return cont(empty());
                }
                break;
            }
        };
        return (matchValue_2 = loop((x) => x), isEmpty(matchValue_2) ? empty() : ((t_2 = tail(matchValue_2), (h_2 = head(matchValue_2), isEmpty(h_2) ? t_2 : cons(h_2, t_2)))));
    }
    finally {
        disposeSafe(en);
    }
}

export function intersect(seq1, seq2) {
    const patternInput = (length_1(seq1) >= length_1(seq2)) ? [seq2, seq1] : [seq1, seq2];
    const hsSs = new HashSet([], HashIdentity_Structural());
    iterate((arg_1) => {
        addToSet(arg_1, hsSs);
    }, patternInput[0]);
    intersectWith(hsSs, patternInput[1]);
    return hsSs;
}

export function groupsOfAtMost(size, s) {
    return delay(() => {
        const en = getEnumerator(s);
        const more = new FSharpRef(true);
        return enumerateWhile(() => more.contents, delay(() => {
            const group = toList(delay(() => {
                const i = new FSharpRef(0);
                return enumerateWhile(() => ((i.contents < size) && en["System.Collections.IEnumerator.MoveNext"]()), delay(() => append(singleton(en["System.Collections.Generic.IEnumerator`1.get_Current"]()), delay(() => {
                    i.contents = ((i.contents + 1) | 0);
                    return empty_1();
                }))));
            }));
            if (isEmpty(group)) {
                more.contents = false;
                return empty_1();
            }
            else {
                return singleton(group);
            }
        }));
    });
}

export function countIf(f, input) {
    const loop = (en_1_mut, counter_mut) => {
        loop:
        while (true) {
            const en_1 = en_1_mut, counter = counter_mut;
            if (en_1["System.Collections.IEnumerator.MoveNext"]()) {
                if (f(en_1["System.Collections.Generic.IEnumerator`1.get_Current"]())) {
                    en_1_mut = en_1;
                    counter_mut = (counter + 1);
                    continue loop;
                }
                else {
                    en_1_mut = en_1;
                    counter_mut = counter;
                    continue loop;
                }
            }
            else {
                return counter | 0;
            }
            break;
        }
    };
    return loop(getEnumerator(input), 0) | 0;
}

export function pivotize(aggregation, defaultValue, keyList, valueList) {
    const m = ofSeq(valueList, {
        Compare: compare,
    });
    return map((k) => {
        if (FSharpMap__ContainsKey(m, k)) {
            return aggregation(FSharpMap__get_Item(m, k));
        }
        else {
            return defaultValue;
        }
    }, keyList);
}

export function tryHead(s) {
    return tryPick(some, s);
}

export function headOrDefault(defaultValue, s) {
    const matchValue = tryHead(s);
    if (matchValue == null) {
        return defaultValue;
    }
    else {
        return value_1(matchValue);
    }
}

export function unzip(input) {
    const patternInput = foldBack((tupledArg, tupledArg_1) => [cons(tupledArg[0], tupledArg_1[0]), cons(tupledArg[1], tupledArg_1[1])], input, [empty(), empty()]);
    return [ofList(patternInput[0]), ofList(patternInput[1])];
}

export function unzip3(input) {
    const patternInput = foldBack((tupledArg, tupledArg_1) => [cons(tupledArg[0], tupledArg_1[0]), cons(tupledArg[1], tupledArg_1[1]), cons(tupledArg[2], tupledArg_1[2])], input, [empty(), empty(), empty()]);
    return [ofList(patternInput[0]), ofList(patternInput[1]), ofList(patternInput[2])];
}

export function foldi(f, acc, sequence) {
    const en = getEnumerator(sequence);
    const loop = (i_mut, acc_1_mut) => {
        loop:
        while (true) {
            const i = i_mut, acc_1 = acc_1_mut;
            const matchValue = en["System.Collections.IEnumerator.MoveNext"]();
            if (matchValue) {
                i_mut = (i + 1);
                acc_1_mut = f(i, acc_1, en["System.Collections.Generic.IEnumerator`1.get_Current"]());
                continue loop;
            }
            else {
                return acc_1;
            }
            break;
        }
    };
    return loop(0, acc);
}

export function countDistinctBy(keyf, sequence) {
    const dict = new Dictionary([], HashIdentity_Structural());
    const en = getEnumerator(sequence);
    while (en["System.Collections.IEnumerator.MoveNext"]()) {
        const key = keyf(en["System.Collections.Generic.IEnumerator`1.get_Current"]());
        let matchValue;
        let outArg = 0;
        matchValue = [tryGetValue(dict, key, new FSharpRef(() => outArg, (v) => {
            outArg = (v | 0);
        })), outArg];
        if (matchValue[0]) {
            dict.set(key, matchValue[1] + 1);
        }
        else {
            dict.set(key, 1);
        }
    }
    return delay(() => map((v_1) => [v_1[0], v_1[1]], dict));
}

export function joinCross(s) {
    return choose((v) => {
        let matchResult, left, right;
        if (v[0] != null) {
            if (v[1] != null) {
                matchResult = 0;
                left = value_1(v[0]);
                right = value_1(v[1]);
            }
            else {
                matchResult = 1;
            }
        }
        else {
            matchResult = 1;
        }
        switch (matchResult) {
            case 0:
                return [left, right];
            default:
                return void 0;
        }
    }, s);
}

export function joinLeft(s) {
    return choose((v) => {
        let matchResult, left, right;
        if (v[0] != null) {
            if (v[1] != null) {
                matchResult = 0;
                left = value_1(v[0]);
                right = value_1(v[1]);
            }
            else {
                matchResult = 1;
            }
        }
        else {
            matchResult = 1;
        }
        switch (matchResult) {
            case 0:
                return [left, right];
            default:
                return void 0;
        }
    }, s);
}

export function joinRight(s) {
    return choose((v) => {
        let matchResult, left, right;
        if (v[0] != null) {
            if (v[1] != null) {
                matchResult = 0;
                left = value_1(v[0]);
                right = value_1(v[1]);
            }
            else {
                matchResult = 1;
            }
        }
        else {
            matchResult = 1;
        }
        switch (matchResult) {
            case 0:
                return [left, right];
            default:
                return void 0;
        }
    }, s);
}

export function joinBy(joinOption, keyf1, keyf2, s1, s2) {
    const dict = new Dictionary([], {
        Equals: equals,
        GetHashCode: structuralHash,
    });
    iterate((l) => {
        const key = keyf1(l);
        const lValue = l;
        let patternInput;
        let outArg = defaultOf();
        patternInput = [tryGetValue(dict, key, new FSharpRef(() => outArg, (v) => {
            outArg = v;
        })), outArg];
        if (patternInput[0]) {
            dict.set(key, [some(lValue), patternInput[1][1]]);
        }
        else {
            addToDict(dict, key, [some(lValue), void 0]);
        }
    }, s1);
    iterate((r) => {
        const key_1 = keyf2(r);
        const rValue = r;
        let patternInput_1;
        let outArg_1 = defaultOf();
        patternInput_1 = [tryGetValue(dict, key_1, new FSharpRef(() => outArg_1, (v_1) => {
            outArg_1 = v_1;
        })), outArg_1];
        if (patternInput_1[0]) {
            dict.set(key_1, [patternInput_1[1][0], some(rValue)]);
        }
        else {
            addToDict(dict, key_1, [void 0, some(rValue)]);
        }
    }, s2);
    return joinOption(map((group) => group[1], dict));
}

export function Double_seqInit(from, tto, length) {
    const stepWidth = (tto - from) / (length - 1);
    return initialize(~~length, (x) => ((x * stepWidth) + from));
}

export function Double_seqInitStepWidth(from, tto, stepWidth) {
    return rangeDouble(from, stepWidth, tto);
}

export function Double_filterNaN(sq) {
    return filter((x) => !Number.isNaN(x), sq);
}

export function Double_filterNanBy(f, sq) {
    return filter((x) => !Number.isNaN(f(x)), sq);
}

export function Double_filterInfinity(sq) {
    return filter((x) => !isInfinity(x), sq);
}

export function Double_filterInfinityBy(f, sq) {
    return filter((x) => !isInfinity(f(x)), sq);
}

export function Double_filterNanAndInfinity(sq) {
    return filter((x) => !(Number.isNaN(x) ? true : isInfinity(x)), sq);
}

export function Double_filterNanAndInfinityBy(f, sq) {
    return filter((v) => {
        const x = f(v);
        return !(Number.isNaN(x) ? true : isInfinity(x));
    }, sq);
}

export function Double_existsNaN(sq) {
    return exists((x) => Number.isNaN(x), sq);
}

export function Double_existsNanBy(f, sq) {
    return exists((x) => Number.isNaN(f(x)), sq);
}

File: src/FSharpAux.Core/Seq.fs
namespace FSharpAux

open System.Collections.Generic

[<AutoOpen>]
module Seq =    
    
    ///Adds a value to the back of a sequence.
    let appendSingleton (s : seq<'T>) (value : 'T) =
        Seq.append s (Seq.singleton value)

    ///Adds a value to the front of a sequence.
    let consSingleton (s : seq<'T>) (value : 'T) =
        Seq.append (Seq.singleton value) s

    /// Initialize a sequence of length and repeated value (like R! repeat : but swapped input).
    let initRepeatValue length value =
        Seq.initInfinite (fun _ -> value) |> Seq.take length

    /// Initialize a sequence of length and repeated values (like R! repeat: but swapped input).
    let initRepeatValues length values =
        Seq.initInfinite (fun _ -> values) |> Seq.take length |> Seq.concat

    /// Iterates over the elements of the input sequence and groups adjacent
    /// elements. A new group is started after the specified predicate holds 
    /// about the element of the sequence (and at the beginning of the iteration).
    // with tail recursion and continuous passing 
    // [3;3;2;4;1;2] |> Seq.groupAfter (fun n -> n%2 = 1);;
    let groupAfter f (input : seq<_>) =     
        let rec group (en : IEnumerator<_>) cont acc c  =            
                if not (f en.Current) && en.MoveNext() then
                    group en (fun l -> cont <| c :: l) acc en.Current
                else
                    (fun l -> cont <| c :: l) []
        seq {
            use en = input.GetEnumerator()
            while en.MoveNext() do
                yield group en id [] en.Current
        }

    /// Iterates over elements of the input sequence and groups adjacent elements.
    /// A new group is started when the specified predicate holds about the element
    /// of the sequence (and at the beginning of the iteration).
    ///
    /// For example: 
    ///    Seq.groupWhen isOdd [3;3;2;4;1;2] = seq [[3]; [3; 2; 4]; [1; 2]]
    let groupWhen f (input : seq<'a>) =
        use en = input.GetEnumerator()
        // in older commits, if the first case returned true, the first element (i.e., the last element of the input seq) was concatinated to the next element instead of standing alone.
        // this is not the most elegant solution, but the only working one I came up with.
        let mutable firstCase = false 
    
        // this function matches the cases and iterats from the last to the first element of the input sequence (tested via print debugging), 
        // concatinating elements if false and concatinating the element with an empty list if true.
        let rec loop cont =
            if en.MoveNext() then
                if (f en.Current) then
                    let temp = en.Current
                    loop (fun y ->
                        cont
                            (
                                match y, firstCase with
                                | _ :: _, true ->
                                    firstCase <- false
                                    [] :: [temp] :: y 
                                | h :: t, false ->
                                    [] :: (temp :: h) :: t 
                                | [], _ -> 
                                    firstCase <- true
                                    [[temp]]
                            )
                    )
                else
                    let temp = en.Current
                    loop (fun y ->
                        cont
                            (
                                match y, firstCase with
                                | _ :: _, true ->
                                    firstCase <- false
                                    [temp] :: y 
                                | h :: t, false ->
                                    (temp :: h) :: t
                                | [], _ -> 
                                    [[temp]]
                            )
                    )
            else
                cont []
        // Remove when first element is empty due to "[] :: (temp :: h) :: t"
        let tmp : seq<seq<'a>> = 
            match (loop id) with
            | h :: t -> 
                match h with
                | [] -> t
                | _  -> h :: t
            | [] -> 
                []
            |> Seq.cast
    
        tmp

        /// Computes the intersection of two sequences.
    let intersect (seq1 : seq<'T>) seq2 : seq<'T> = 
        let smallerSeq, largerSeq =
            if Seq.length seq1 >= Seq.length seq2 then seq2, seq1
            else seq1, seq2
        let hsSs = HashSet<'T>(HashIdentity.Structural<'T>)
        smallerSeq |> Seq.iter (hsSs.Add >> ignore)
        hsSs.IntersectWith largerSeq
        hsSs

    /// <summary>
    /// Divides the input sequence into chunks depending on the projection function.
    /// 
    /// A new chunk is created each time the projection function returns true.
    ///
    /// Example: Seq.chunkBy isOdd [3;3;2;4;1;2] = seq [[3]; [3; 2; 4]; [1; 2]]
    /// </summary>
    /// <param name="projection">The function to determine if an element creates a new chunk</param>
    /// <param name="source"></param>
    let chunkBy (projection: 'T -> 'Key) (source: seq<'T>) = 
        seq {
            use e = source.GetEnumerator ()
            if e.MoveNext () then
                let mutable g = projection e.Current
                let mutable members = ResizeArray ()
                members.Add e.Current
                while e.MoveNext () do
                    let key = projection e.Current
                    if g = key then 
                        members.Add e.Current
                    else
                        yield g, members |> Seq.cast<'T>
                        g <- key
                        members <- ResizeArray ()
                        members.Add e.Current
                yield g, members |> Seq.cast<'T>
        }

// // Without continuation passing

//    let groupWhen f (input:seq<_>) = seq {
//        use en = input.GetEnumerator()
//        let running = ref true
//    
//        // Generate a group starting with the current element. Stops generating
//        // when it founds element such that 'f en.Current' is 'true'
//        let rec group() = 
//            [ yield en.Current
//              if en.MoveNext() then
//                if not (f en.Current) then yield! group() 
//              else running := false ]
//    
//        if en.MoveNext() then
//            // While there are still elements, start a new group
//            while running.Value do
//            yield group() |> Seq.ofList }





    
    /// Break sequence into n-element subsequences.
    let groupsOfAtMost (size : int) (s : seq<'v>) : seq<list<'v>> =
        seq {
            let en = s.GetEnumerator()
            let more = ref true
            while !more do
            let group =
                [
                    let i = ref 0
                    while !i < size && en.MoveNext() do
                        yield en.Current
                        i := !i + 1
                ]
            if List.isEmpty group then
                more := false
            else
                yield group
        }


    /// Iterates over elements of the input sequence and increase the counter
    /// if the function returens true.
    let countIf f (input:seq<_>) =         
        let en = input.GetEnumerator()
        let rec loop (en: IEnumerator<_>) (counter:int) =
            if en.MoveNext() then
                if (f en.Current) then
                    loop en (counter + 1)
                else
                    loop en counter                    
            else
                counter
        loop en 0


    let pivotize (aggregation:seq<'T> -> 'A) (defaultValue:'A) (keyList:seq<'key>) (valueList:seq<'key*seq<'T>>) =
        let m = valueList |> Map.ofSeq
        keyList 
        |> Seq.map (
            fun k -> 
                match m.TryGetValue k with
                | true, mk ->
                    aggregation mk
                | false, _ ->
                    defaultValue 
        )


    // /// Combines two sequences according to key generating functions.
    //let joinBy (keyf1: 'a -> 'key) (keyf2: 'b -> 'key) (s1:seq<'a>) (s2:seq<'b>) =
    //    // Wrap a StructBox(_) around all keys in case the key type is itself a type using null as a representation
    //        let dict = new Dictionary<'key,'a option*'b option>()

    //        let insertLeft key lValue =    
    //            let ok,refV = dict.TryGetValue(key)
    //            if ok then 
    //                let _,b = refV
    //                dict.[key] <- (Some lValue,b)            
    //            else             
    //                dict.Add(key,(Some lValue,None))
            
            
    //        let insertRight key rValue =    
    //            let ok,refV = dict.TryGetValue(key)
    //            if ok then 
    //                let a,_ = refV
    //                dict.[key] <- (a,Some rValue) 
    //            else             
    //                dict.Add(key,(None,Some rValue))
            

    //        s1 |> Seq.iter (fun l -> insertLeft  (keyf1 l) l)                                         
    //        s2 |> Seq.iter (fun r -> insertRight (keyf2 r) r)
    
    //        dict |> Seq.map (fun group -> group.Value)
                        


   
    
    /// Returns head of a seq as option or None if seq is empty.
    let tryHead s = Seq.tryPick Some s
    
    /// Returns head of a seq or default value if seq is empty.
    let headOrDefault defaultValue s =
        match (tryHead s) with
        | Some x    -> x
        | None      -> defaultValue

    /// Splits a sequence of pairs into two sequences.
    let unzip (input : seq<_>) =
        let (lstA, lstB) = 
            Seq.foldBack (fun (a,b) (accA, accB) -> 
                a :: accA, b :: accB) input ([], [])
        (Seq.ofList lstA, Seq.ofList lstB)    

    /// Splits a sequence of triples into three sequences.
    let unzip3 (input : seq<_>) =
        let (lstA, lstB, lstC) = 
            Seq.foldBack (fun (a,b,c) (accA, accB, accC) -> 
                a :: accA, b :: accB, c :: accC) input ([],[],[])
        (Seq.ofList lstA, Seq.ofList lstB, Seq.ofList lstC)

    let foldi (f : int -> 'State -> 'T -> 'State) (acc : 'State) (sequence : seq<'T>) =
        let en = sequence.GetEnumerator()
        let rec loop i acc = 
            match en.MoveNext() with
            | false -> acc
            | true -> loop (i + 1) (f i acc en.Current)
        loop 0 acc

    /// Applies a keyfunction to each element and counts the amount of each distinct resulting key.
    let countDistinctBy (keyf : 'T -> 'Key) (sequence : seq<'T>) =
        let dict = Dictionary<_, int> HashIdentity.Structural<'Key>
        let en = sequence.GetEnumerator()
        // Build the distinct-key dictionary with count
        do 
            while en.MoveNext() do
                let key = keyf en.Current
                match dict.TryGetValue(key) with
                | true, count ->
                        dict.[key] <- count + 1 // If it matches a key in the dictionary increment by one
                | _ -> 
                        dict.[key] <- 1 // If it doesn't match create a new count for this key
        // Write to Sequence
        seq {for v in dict do yield v.Key |> Operators.id, v.Value}
    
    
    type JoinOption<'a, 'b, 'c> = seq<'a option * 'b option> -> seq<'c>

    let joinCross (s : seq<'a option * 'b option>) =
        s
        |> Seq.choose (
            fun v -> 
                match v with
                | (Some left, Some right)   -> Some (left,right)
                | _,_                       -> None
        )

    let joinLeft (s : seq<'a option * 'b option>) =
        s
        |> Seq.choose (
            fun v -> 
                match v with
                | (Some left, Some right)   -> Some (left,right)
                | _,_                       -> None
        )

    let joinRight (s : seq<'a option * 'b option>) =
        s
        |> Seq.choose (
            fun v -> 
                match v with
                | (Some left, Some right)   -> Some (left,right)
                | _,_                       -> None
        )

    // Combines two sequences according to key generating functions.
    let joinBy (joinOption:JoinOption<'a,'b, 'c>) (keyf1: 'a -> 'key) (keyf2: 'b -> 'key) (s1:seq<'a>) (s2:seq<'b>) =
        // Wrap a StructBox(_) around all keys in case the key type is itself a type using null as a representation
            let dict = new Dictionary<'key, 'a option * 'b option>()

            let insertLeft key lValue =    
                let ok,refV = dict.TryGetValue(key)
                if ok then 
                    let _,b = refV
                    dict.[key] <- (Some lValue, b)            
                else             
                    dict.Add(key, (Some lValue, None))
            
            
            let insertRight key rValue =    
                let ok,refV = dict.TryGetValue(key)
                if ok then 
                    let a,_ = refV
                    dict.[key] <- (a, Some rValue) 
                else             
                    dict.Add(key, (None, Some rValue))
            

            s1 |> Seq.iter (fun l -> insertLeft  (keyf1 l) l)                                         
            s2 |> Seq.iter (fun r -> insertRight (keyf2 r) r)
    
            dict 
            |> Seq.map (fun group -> group.Value)
            |> joinOption           


    // #region seq double extension
    
    /// Seq module extensions specialized for seq<float>.
    module Double = 

        /// Generates a sequence (like R! seq.int).
        let seqInit (from : float) (tto : float) (length : float) =
            let stepWidth = (tto - from) / (length - 1.)
            Seq.init (int length) ( fun x -> (float x * stepWidth) + from)    

        /// Generates sequence given step width (like R! seq).
        let seqInitStepWidth (from : float) (tto : float) (stepWidth : float) =
            seq { from .. stepWidth .. tto }

        /// Returns the copy of an input sequence where nans were filtered out.
        let filterNaN (sq : seq<float>) =
            sq |> Seq.filter (fun x -> not (System.Double.IsNaN(x)))

        /// Returns the copy of an input sequence where nans were filtered out after applying f to every element.
        let filterNanBy (f : 'a -> float) (sq : seq<'a>) =
            sq |> Seq.filter (fun x -> not (System.Double.IsNaN(f x)))

        /// Returns the copy of an input sequence where infinitys were filtered out.
        let filterInfinity (sq : seq<float>) =
            sq |> Seq.filter (fun x -> not (System.Double.IsInfinity(x)))
    
        /// /// Returns the copy of an input sequence where infinitys were filtered out after applying f to every element.
        let filterInfinityBy (f : 'a -> float) (sq : seq<'a>) =
            sq |> Seq.filter (fun x -> not (System.Double.IsInfinity(f x)))

        /// Returns the copy of an input sequence where nans and infinitys were filtered out.
        let filterNanAndInfinity (sq : seq<float>) =
            sq |> Seq.filter (fun x -> not (System.Double.IsNaN(x) || System.Double.IsInfinity(x)))

        /// Returns the copy of an input sequence where nans and infinitys were filtered out after applying f to every element.
        let filterNanAndInfinityBy (f : 'a -> float) (sq : seq<'a>) =
            sq 
            |> Seq.filter (
                fun v ->
                    let x = f v
                    not (System.Double.IsNaN(x) || System.Double.IsInfinity(x))
            )

        /// Returns true if the input sequence contains nan.
        let existsNaN (sq : seq<float>) =
            sq |> Seq.exists (fun x -> System.Double.IsNaN(x))

        /// Returns true if the input sequence contains nan after applying f to every element.
        let existsNanBy (f : 'a -> float) (sq : seq<'a>) =
            sq |> Seq.exists (fun x -> System.Double.IsNaN(f x))


    //#endregion seq double extension

    
